text/3680-use.md

```diff
@@ -0,0 +1,457 @@
+
+`.use` supports chaining, so in particular it works when calling a method that
+would otherwise consume `self`:
+
+```rust
+obj.use.consume();
+obj.method();
+```
+
+Calling `x.use` requires that the type of `x` implement the `Use` trait. This
+trait identifies types whose clone implementation is lightweight, such as
+reference-counted types.
+
+Various types in the standard library implement `Use`. You can implement this
+trait for your own types, if they meet the requirements for being lightweight
+to clone. (See the [reference-level explanation][reference-level-explanation]
+for the requirements.)
+
+```rust
+impl Use for MyType {}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## The `Use` trait
+
+```rust
+/// Trait for objects whose clone impl is lightweight (e.g. reference-counted)
+///
+/// Cloning an object implementing this trait should in general:
+/// - be O(1) (constant) time regardless of the amount of data managed by the object,
+/// - not require a memory allocation,
+/// - not require copying more than roughly 64 bytes (a typical cache line size),
+/// - not block,
+/// - not have any semantic side effects (e.g. allocating a file descriptor), and
+/// - not have overhead larger than a couple of atomic operations.
+///
+/// The `Use` trait does not provide a method; instead, it indicates that
+/// `Clone::clone` is lightweight, and allows the use of the `.use` syntax.
+trait Use: Clone {}
+```
+
+This trait should be implemented for anything in the standard library that
+meets these criteria. Some notable types in the standard library that implement
+`Use`:
+
+- `std::sync::Arc`
+- `std::sync::Weak`
+- `std::rc::Rc`
+- `std::rc::Weak`
+- `std::sync::mpsc::Sender` and `std::sync::mpsc::SyncSender`
+- Tuples of types whose components all implement `Use`
+- `Option<T>` where `T: Use`
+- `Result<T, E>` where `T: Use` and `E: Use`
+
+Some notable types that implement `Clone` but should *not* implement `Use`:
+arrays, `String`, `Box`, `Vec`, `HashMap`, and `BTreeMap`.
+
+We may want to add a clippy or rustc lint (e.g. `expensive_use`) for
+implementations of `Use` on an excessively large type, or a type whose `clone`
+implementation seems to be obviously breaking the intended constraints on the
+`Use` trait.  Such a lint would be best-effort only, and could always be marked
+as `allow` by a crate, but could help to discourage such implementations.
+
+We may want to add a clippy or rustc lint for calls to `.clone()` that could
+use `.use` instead. This would help the remaining calls to `.clone()` stand out
+as "expensive" clones. (Such a lint would need to take MSRV into account before
+making such a suggestion; clippy already has such a mechanism and rustc may
+gain one in the future.)
+
+We may want to add a `derive(Use)`, which automatically adds `Use` bounds on
+fields just as `derive(Clone)` adds `Clone` bounds on fields.
+
+## The implementation and optimization of `.use`
```

Comment:
This section to me seems the most out-there part of the RFC, not because I think that it's a bad idea (I like automatically optimizing out ref-counting) but because none of these optimisations are apparently possible from the original trait. Specifically:

* It's not clear whether "not cloning" here means being able to pass `&T`, or `&Arc<T>`, or both. Note that `&Arc<T>` *feels* like a useless type on the surface but is common in cases where you may want to keep your own copy of the data, but need to inspect it in order to decide.
* The interchangeability of `Arc<T>`, `&Arc<T>`, and `&T` here really specifically applies to "ref-tracking" types specifically, and that should really be represented in the parent trait via an `AsRef` or `Deref` bound specifically.

To me, it feels like this RFC is trying to overgeneralise to something beyond its original use case, which is to make "reference-tracked" (ref-counted, or even GC'd) types easier to manage in Rust. So, I think that it should try to narrow down more specifically on what those types look like and maybe use a different syntax (`.ref` feels weird but I could probably get used to it) to help facilitate that.

Reply:
Also adding onto this, I really do want to press more on the idea of "lightweight clones that aren't ref-tracking"

Like, the only example in the RFC text is `Sender`, which still seems to feel like a ref-tracking type. You have multiple different references to the channel that the OS is tracking by keeping the receiver open until they all go away.

Stuff like `Range` really should be `Copy` and eventually will be.

What else is left?

Reply:
well, a wrapper over `slice::Iter` for a byte parser would be lightweight to clone but not `Copy`...or just look at almost any `<&T as IntoIterator>::IntoIter` type

Reply:
They are lightweight clones, but the real question is whether they would benefit substantially from this extra syntax. I think that the kinds of "cloneable" things most suitable for this kind of syntax are the immutable reference kind, and not anything that is cheap to clone but encapsulates nontrivial state.

Note that I'm counting the references as immutable, not the actual underlying data. Mutating an iterator and cloning it are not reorderable operations, whereas this is the case for a smart pointer with interior mutability. Since this API talks a lot about whether clones can be elided or not, I think that it would be useful to preserve the property that clones can also be reordered among other operations, because otherwise many optimisations that would be possible are no longer possible.